core/src/Cuda/Kokkos_Cuda_Instance.cpp

```diff
@@ -278,10 +279,13 @@ void CudaInternal::initialize(cudaStream_t stream) {
 
   // Allocate a staging buffer for constant mem in pinned host memory
   // and an event to avoid overwriting driver for previous kernel launches
-  if (!constantMemHostStagingPerDevice[m_cudaDev])
+  if (!constantMemHostStagingPerDevice[m_cudaDev]) {
+    void *constant_memory_void_ptr = nullptr;
     KOKKOS_IMPL_CUDA_SAFE_CALL((cuda_malloc_host_wrapper(
-        reinterpret_cast<void **>(&constantMemHostStagingPerDevice[m_cudaDev]),
-        CudaTraits::ConstantMemoryUsage)));
+        &constant_memory_void_ptr, CudaTraits::ConstantMemoryUsage)));
+    constantMemHostStagingPerDevice[m_cudaDev] =
+        static_cast<unsigned long *>(constant_memory_void_ptr);
```

Comment:
I really dislike the spelling `unsigned long` here.
In contrast to the other changes in that file one cannot tell it is right from the context in that file without jumping to definition twice in other files.
I don't see this as a good change.

Reply:
Is 

```c++
static_cast<decltype(constantMemHostStagingPerDevice)::mapped_type>(...
```

better? Or maybe alias the `std::map<int, unsigned long *>` in CudaInternal and use that?

Reply:
The pointer assignment would fail if the types were different anyway (and none of them is `void*`).

Reply:
> The pointer assignment would fail if the types were different anyway (and none of them is `void*`).

Sure, I get that the compiler would tell you, I am saying I cannot say this is fine when reading the code.
Per my question on the Slack developer channel, if reviewers confirm that they don't see it as an issue I will merge.